compiler/AST/AstToText.cpp

```diff
@@ -896,7 +898,139 @@ static bool looksLikeInfixOperator(const char *fnName)
+    return 13;
+  else if (unary &&
+           (strcmp(op, "+") == 0 || strcmp(op, "-") == 0))
+    return 12;
+  else if (strcmp(op, "<<") == 0 || strcmp(op, ">>") == 0)
+    return 11;
+  else if (strcmp(op, "&") == 0)
+    return 10;
+  else if (strcmp(op, "^") == 0)
+    return 9;
+  else if (strcmp(op, "|") == 0)
+    return 8;
+  else if (strcmp(op, "+") == 0 || strcmp(op, "-") == 0)
+    return 7;
+  // .. and ..< are precedence 6, but don't come through this path.
+  else if (strcmp(op, "<") == 0 || strcmp(op, "<=") == 0 ||
+           strcmp(op, ">") == 0 || strcmp(op, ">=") == 0)
+    return 5;
+  else if (strcmp(op, "==") == 0 || strcmp(op, "!=") == 0)
+    return 4;
+  else if (strcmp(op, "&&") == 0)
+    return 3;
+  else if (strcmp(op, "||") == 0)
+    return 2;
+  // by and align are precedence 1 too, but don't come through this path.
+  else if (strcmp(op, "#") == 0)
+    return 1;
+  printf("unknown op %s", op);
+  return 0;
+}
+
+/*
+ * needParens(outer, inner, ...) is called to evaluate whether we need
+ * to add parens around the inner op.  We're here when
+ * appendExpr(outer) is calling appendExpr(expr->get(1)) or
+ * expr->get(2), which has become appendExpr(inner).
+ *
+ * Given an AST node outer, with one child inner, this function
+ * returns true if we need to emit parens around the child expression.
+ * That is, if emitting the expression without parenthesis would
+ * change the semantics from what the AST represents.
+ *
+ * If the inner (child) operator has higher precedence than the outer
+ * (parent), then we don't need parens, as emitting
+ * "a outer_op b inner_op c" is equivalent to
+ * "a outer_op (b inner_op c)".
+ *
+ * If the child operator has equal precedence to the outer, then we
+ * generally don't need parenthesis, except for a few exceptions.  If
+ * the desired expression is a-(b-c) or a-(b+c), then the parentheses
+ * are needed.  Of course they're not needed for a+(b-c) or a+(b+c).
+ * If the desired expression is a/(b/c) or a/(b*c) or a%(b/c), they're
+ * needed, but not for a*(b*c) or a*(b/c).
+ * (TODO: a*(b/c) might need the parens for overflow reasons.)
+ *
+ * Unary - (and +) have higher precedence than binary - (and +), so we
+ * need the unary flag to know which case we're in.
+ *
+ * The postfix flag is needed because postfix ! has higher precedence
+ * than prefix !.
+ *
+ * The innerIsRHS flag is needed because we need parens to express
+ * a-(b-c) where inner- is the RHS of the outer-.  But we can emit
+ * a-b-c instead of (a-b)-c -- when inner- is the LHS of outer-.
+ */
+static bool needParens(const char *outer, const char *inner,
+                       bool outerUnary, bool outerPostfix, bool innerIsRHS) {
+  bool ret = false;
+  int outerprec, innerprec;
+
+  if (!outer)
+    return false;
+
+  outerprec = opToPrecedence(outer, outerUnary, outerPostfix);
+  innerprec = opToPrecedence(inner, false, false); // inner is a binary op
```

Comment:
Is that a fair assumption?  I'd like to see a test case with a unary operator on the right, e.g. `a + (-b)` or something.  Or maybe `a && !b` or `a && ! (b || c)` would make more sense

Reply:
FWIW, with the unnecessary parens approach, I was able to see the difference between `fnName` of `:` vs `_cast`. 

For `var unarymcast = -a: uint` we get `inner == _cast`, `outer == -`.
For `var Unarymcast = (-a):uint;` we get `outer == :`, `inner == -`.

That seems weird.  The innerness/outerness of the cast makes sense in both, but I didn't expect different `fnName`s for the cast.

Reply:
Now that it's merged to master, the cast operator shows up as `:` in both cases.  I presume that's due to catching up with `operator :`, but I'm not going to look into it further to check.